src/optimizer/join_elimination.cpp

```diff
@@ -0,0 +1,148 @@
+			all_outer_bindings.insert(binding);
+		}
+		for (auto &binding : unique_column_references) {
+			if (all_outer_bindings.find(binding) == all_outer_bindings.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+
+		if (all_matched) {
+			op = std::move(outer_child);
+		}
+
+		break;
+	}
+	case LogicalOperatorType::LOGICAL_AGGREGATE_AND_GROUP_BY: {
+		auto &aggr = op->Cast<LogicalAggregate>();
+		for (idx_t col_idx = 0; col_idx < aggr.expressions.size(); col_idx++) {
+			if (aggr.expressions[col_idx]->GetExpressionClass() != ExpressionClass::BOUND_AGGREGATE) {
+				continue;
+			}
+			auto &aggr_expr = aggr.expressions[col_idx]->Cast<BoundAggregateExpression>();
+			// Collect the unique columns.
+			bool has_unique_column = false;
+			if (aggr_expr.IsDistinct()) {
+				has_unique_column = true;
+			} else {
+				auto agg_expr_function_name = aggr_expr.function.name;
+				if (agg_expr_function_name == "max" || agg_expr_function_name == "min") {
+					has_unique_column = true;
+				}
+			}
+
+			if (has_unique_column) {
+				for (auto &child : aggr_expr.children) {
+					if (child->expression_class == ExpressionClass::BOUND_COLUMN_REF) {
+						unique_column_references.insert(child->Cast<BoundColumnRefExpression>().binding);
+					}
+				}
+			}
+		}
+		break;
+	}
+	case LogicalOperatorType::LOGICAL_DISTINCT: {
+		auto &distinct = op->Cast<LogicalDistinct>();
+		for (idx_t col_idx = 0; col_idx < distinct.distinct_targets.size(); col_idx++) {
+			if (distinct.distinct_targets[col_idx]->GetExpressionClass() != ExpressionClass::BOUND_COLUMN_REF) {
+				continue;
+			}
+			unique_column_references.insert(
+			    distinct.distinct_targets[col_idx]->Cast<BoundColumnRefExpression>().binding);
+		}
+		break;
+	}
+	default:
+		break;
+	}
+
+	LogicalOperatorVisitor::VisitOperatorExpressions(*op);
+
+	column_binding_set_t parent_column_references;
+	column_binding_set_t parent_unique_column_references;
+	if (op->children.size() > 1) {
+		// If there are more than one children. We should copy the column reference from the father
+		// and not let each child influence each other.
+		parent_column_references = column_references;
+		parent_unique_column_references = unique_column_references;
+	}
+	for (idx_t idx = 0; idx < op->children.size(); idx++) {
+		auto &child = op->children[idx];
+		if (idx > 0) {
+			D_ASSERT(idx < 2);
+			// If we can guarantee that the number of children of all operators are less equal than 2.
+			// We can use 'std::move' to save the copy cost.
+			column_references = std::move(parent_column_references);
+			unique_column_references = std::move(parent_unique_column_references);
+		}
+		child = Optimize(std::move(child));
+	}
+	return op;
+}
+
+unique_ptr<Expression> JoinElimination::VisitReplace(BoundColumnRefExpression &expr, unique_ptr<Expression> *expr_ptr) {
+	// Add a column reference to record which column has been used.
+	column_references.insert(expr.binding);
+	return nullptr;
+}
```

Comment:
@Mytherin Another question is here. Now we can use this function to handle all the `BoundColumnRefExpression` in all of the expressions in this plan tree. But now, we only can do one operation like `column_references.insert(expr.binding);`. So If I want to do different operations in different places, How can I do it? Because maybe there are a lot of places where I want to extract the column references from the expressions and do some specific operations.